File: Sources/ExperimentKit/FreezePolicy.swift
import Foundation
import SteeringKit

/// Gate decisions over a manifest and supplied evidence values. This owner
/// never reads files, chooses a workspace, loads a model or performs a commit.
enum FreezePolicy {
    typealias Outcome = ExperimentStore.FreezeGateOutcome

    struct ValidationFacts {
        let hasMatchingEvidence: Bool
        var vacuousProblem: String? = nil
        var vacuousRepair: String? = nil
    }

    struct BatteryFacts {
        let hasMatchingEvidence: Bool
        var results: [CapabilityBatteryConditionResult] = []
        var expectedHash: String? = nil
    }

    static func freezeRefusal(_ failures: [ExperimentStore.FreezeGateOutcome]) -> Error? {
        guard let first = failures.first else { return nil }
        if let underlying = first.underlying { return underlying }
        let failed = Set(failures.map(\.gate))
        return ExperimentError(
            refusal: FreezeRefusal(
                gate: first.gate,
                gates: FreezeGate.allCases.filter(failed.contains),
                reason: first.refusal, repairAction: first.repairAction))
    }

    static func checkedOutcome(
        _ gate: FreezeGate, name: String, repairAction: String,
        result: Result<Void, Error>
    ) -> Outcome? {
        do {
            try result.get()
            return nil
        } catch let error as ExperimentError {
            return ExperimentStore.FreezeGateOutcome(
                gate: gate, refusal: error.reason,
                forced: ManifestDeclarationPolicy.plainGateReason(
                    error.reason, experimentName: name),
                repairAction: repairAction)
        } catch {
            return ExperimentStore.FreezeGateOutcome(
                gate: gate, refusal: "\(error)", forced: "\(error)",
                repairAction: repairAction, underlying: error)
        }
    }

    static func revision(_ manifest: ExperimentManifest, name: String) -> Outcome? {
        guard manifest.modelRevision == nil else { return nil }
        let forced =
            "model revision is not pinned and \(manifest.modelID) is not in "
            + "the local HF cache"
        let repair = "create with --revision, load the model once, or freeze --force"
        return ExperimentStore.FreezeGateOutcome(
            gate: .revision,
            refusal: "cannot freeze '\(name)': \(forced) — \(repair)",
            forced: forced, repairAction: repair)
    }

    static func symbolicRevision(_ manifest: ExperimentManifest, name: String) -> Outcome? {
        guard let symbolic = ManifestDeclarationPolicy.symbolicRevisionProblem(manifest) else {
            return nil
        }
        return ExperimentStore.FreezeGateOutcome(
            gate: .revision, refusal: "cannot freeze '\(name)': \(symbolic)",
            forced: symbolic,
            repairAction: "pin the immutable commit the symbolic revision "
                + "resolves to, or freeze --force")
    }

    static func dtype(_ manifest: ExperimentManifest, name: String) -> Outcome? {
        guard let badDtype = ManifestDeclarationPolicy.unloadableStudyDtypeProblem(manifest) else {
            return nil
        }
        return ExperimentStore.FreezeGateOutcome(
            gate: .measurementPins, refusal: "cannot freeze '\(name)': \(badDtype)",
            forced: badDtype,
            repairAction: "repoint the study dtype at a loadable value, "
                + "or freeze --force")
    }

    static func validation(
        _ manifest: ExperimentManifest, name: String, runSubstrate: String,
        facts: ValidationFacts
    ) -> Outcome? {
        guard ManifestDeclarationPolicy.modelOutputSurfacesOperative(manifest) else { return nil }
        let usesLegacyConceptVectors =
            !manifest.concepts.isEmpty || !manifest.conditions.isEmpty
        guard usesLegacyConceptVectors,
            !ManifestDeclarationPolicy.optvecExemptFromValidateGate(manifest)
        else { return nil }
        guard
            facts.hasMatchingEvidence
        else {
            let forced =
                "no validate run matches its exact pins (model+revision, "
                + "concepts, neutral corpus) on the run substrate " + runSubstrate
            // The engine named is the one whose evidence this gate
            // reads — `steerlab-cli` cannot satisfy a server-substrate
            // gate (gate-5 dry run #2, P2). Sentence structure is
            // unchanged; only the binary moves.
            let repair =
                "Run '\(ManifestDeclarationPolicy.validateCLI(forRunSubstrate: runSubstrate)) "
                + "experiment validate \(name)' first, or "
                + "freeze --force to record an unvalidated experiment"
            return ExperimentStore.FreezeGateOutcome(
                gate: .validateEvidence,
                refusal: "cannot freeze '\(name)': \(forced). \(repair)",
                forced: forced, repairAction: repair)
        }
        // Evidence EXISTS but probed nothing: same gate id, a remedy
        // naming the missing files (2026-08-17).
        guard
            let vacuous = facts.vacuousProblem
        else { return nil }
        // P5 (dry run #1): the gate's own repair FAILED as given.
        // Authoring the named validation.jsonl makes it appear after
        // an attach that pinned it absent, which is a verify()
        // violation — so the very next `validate` refuses. The
        // machine repair now carries the re-attach that re-pins it;
        // the composed PROSE is unchanged (it is the cross-engine
        // refusal string, asserted whole by VacuousValidationTests).
        let repair = facts.vacuousRepair ?? vacuous
        return ExperimentStore.FreezeGateOutcome(
            gate: .validateEvidence, refusal: "cannot freeze '\(name)': \(vacuous)",
            forced: vacuous, repairAction: repair)
    }

    static func checkVariantBatteryEvidence(
        _ manifest: ExperimentManifest, facts: BatteryFacts
    ) throws {
        guard !manifest.variantConditions.isEmpty else { return }
        let name = manifest.name
        guard
            facts.hasMatchingEvidence
        else {
            throw ExperimentError(
                reason: "cannot freeze '\(name)': no validate run matches its exact pins "
                    + "(model+revision, variant capability battery). Variant studies "
                    + "validate the pinned battery per condition — run 'steerlab-cli "
                    + "experiment validate \(name)' first, or freeze --force")
        }
        let results = Dictionary(
            facts.results
                .map { ($0.condition, $0) },
            uniquingKeysWith: { first, _ in first })
        // Forward-referenced conditions (stage 4) are exempt: their agent
        // does not exist at validate time — their battery evidence is the
        // RUN's per-condition battery, produced after server-side
        // resolution.
        let required =
            ["baseline"]
            + manifest.variantConditions
            .filter { $0.fromPromotion == nil }
            .map(\.name)
        let missing = required.filter { results[$0] == nil }
        guard missing.isEmpty else {
            throw ExperimentError(
                reason: "cannot freeze '\(name)': matching validate evidence has no "
                    + "capability-battery results for condition(s): "
                    + missing.joined(separator: ", ")
                    + " — re-run 'steerlab-cli experiment validate \(name)' (each "
                    + "variant condition runs the pinned battery), or freeze --force")
        }
        if let expected = facts.expectedHash {
            let drifted =
                required
                .filter { results[$0]?.batteryHash != expected }
                .sorted()
            guard drifted.isEmpty else {
                throw ExperimentError(
                    reason: "cannot freeze '\(name)': capability battery drifted since "
                        + "validation for condition(s): " + drifted.joined(separator: ", ")
                        + " — re-run 'steerlab-cli experiment validate \(name)', or "
                        + "freeze --force")
            }
        }
    }

    static func checkVariantValidity(
        _ manifest: ExperimentManifest, evidenceGradeVariants: Set<Int>
    ) throws {
        let name = manifest.name
        for (index, variant) in manifest.variantConditions.enumerated() {
            if variant.fromPromotion != nil {
                // Forward-referenced (stage 4): the artifact does not exist
                // at freeze time BY DESIGN — its pins land at run time on
                // the server and are recorded in the run directory.
                // verify() enforces the declaration shape.
                continue
            }
            if variant.artifactHash.isEmpty {
                throw ExperimentError(
                    reason: "cannot freeze '\(name)': variant '\(variant.name)' has no "
                        + "pinned artifactHash")
            }
            for adapter in variant.artifact.adapters
            where (adapter.adapterHash ?? "").isEmpty {
                throw ExperimentError(
                    reason: "cannot freeze '\(name)': variant '\(variant.name)' adapter "
                        + "'\(adapter.name)' has no adapterHash — re-save the variant "
                        + "with hashed adapter weights, or freeze --force")
            }
            let systemPrompt = (variant.artifact.systemPrompt ?? "")
                .trimmingCharacters(in: .whitespacesAndNewlines)
            if !systemPrompt.isEmpty,
                (variant.artifact.systemPromptHash ?? "").isEmpty
            {
                throw ExperimentError(
                    reason: "cannot freeze '\(name)': variant '\(variant.name)' has a "
                        + "system prompt but no systemPromptHash — re-save the variant, "
                        + "or freeze --force")
            }
            for injection in variant.artifact.injections
            where injection.vectorArtifactID.isEmpty {
                throw ExperimentError(
                    reason: "cannot freeze '\(name)': variant '\(variant.name)' has an "
                        + "injection for '\(injection.concept)' without a "
                        + "vectorArtifactID pin")
            }
            // Trained-adapter arms owe the same story about their TRAINING
            // DATA (LoRA readiness §0 amendment 1). An evidence-grade adapter
            // whose dataset is not pinned into the manifest is unverifiable
            // the moment it freezes: the training files could change
            // afterwards with nothing to flag the drift. Exploratory adapters
            // are legal and produce an advisory instead, never a refusal.
            if !variant.artifact.adapters.isEmpty,
                evidenceGradeVariants.contains(index),
                (variant.trainingProvenance?.datasetManifestHash ?? "")
                    .trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
            {
                throw ExperimentError(
                    reason: "cannot freeze '\(name)': variant '\(variant.name)' uses an "
                        + "evidence-grade adapter but carries no "
                        + "trainingProvenance.datasetManifestHash — its training data "
                        + "would stay outside the freeze pin surface. Re-attach the "
                        + "variant so freeze can pin the dataset manifest from the "
                        + "adapter's sidecar, or freeze --force")
            }
        }
    }
}

File: Server/steerlab_server/experiment/freeze_policy.py
"""Freeze gate decisions over a manifest and already-collected evidence.

No paths, file reads, model loads or store imports belong here. Evidence readers
preserve their own integrity checks; this policy owns gate scope and ordering.
"""
from __future__ import annotations
from dataclasses import dataclass
from . import manifest_declaration_policy as declarations
from .manifest_errors import ExperimentStoreError

FORCED_GATE_IDS = ("revision", "validateEvidence", "batteryEvidence",
                   "judgeValidity", "variantValidity", "gitClean",
                   "measurementPins")

@dataclass(frozen=True)
class FreezeEvidence:
    jlens: str | None = None
    variant: str | None = None
    judge: str | None = None
    validation_present: bool = False
    vacuous_validation: str | None = None
    battery: str | None = None
    git: str | None = None


def needs_validation(d: dict) -> bool:
    return bool(declarations.model_output_surfaces_operative(d)
                and (d.get("concepts") or d.get("conditions") or d.get("variantConditions"))
                and not declarations.optvec_exempt_from_validate_gate(d))


def evaluate(name: str, d: dict, evidence: FreezeEvidence) -> list[tuple[str, str]]:
    failures: list[tuple[str, str]] = []
    if not d.get("modelRevision"):
        failures.append(("revision",
            f"cannot freeze '{name}': model revision not pinned and "
            f"{d['modelID']} not in the local HF cache — load it once or freeze --force"))
    symbolic = declarations.symbolic_revision_problem(d)
    if symbolic:
        failures.append(("revision", f"cannot freeze '{name}': {symbolic}"))
    bad_dtype = declarations.unloadable_study_dtype_problem(d)
    if bad_dtype:
        failures.append(("measurementPins", f"cannot freeze '{name}': {bad_dtype}"))
    operative = declarations.model_output_surfaces_operative(d)
    if operative:
        if evidence.jlens is not None:
            failures.append(("measurementPins", evidence.jlens))
        if evidence.variant is not None:
            failures.append(("variantValidity", evidence.variant))
    if evidence.judge is not None:
        failures.append(("judgeValidity", evidence.judge))
    if needs_validation(d):
        if not evidence.validation_present:
            failures.append(("validateEvidence",
                f"cannot freeze '{name}': no validate run matches its exact "
                f"pins — run 'steerlab-server experiment validate {name}' first, or force-freeze"))
        elif evidence.vacuous_validation:
            failures.append(("validateEvidence", evidence.vacuous_validation))
    if operative and d.get("variantConditions") and evidence.battery is not None:
        failures.append(("batteryEvidence", evidence.battery))
    if evidence.git is not None:
        failures.append(("gitClean", evidence.git))
    return failures


def freeze_gate_repair(gate: str, name: str) -> str:
    """The RUNNABLE repair for a freeze-gate refusal on THIS engine.

    Gate-5 dry run #2 (P2/P3): every freeze-gate refusal here carried one
    boilerplate string — "satisfy the named gate, or freeze --force to record
    an explicitly non-citable experiment" — which names no command, and whose
    only concrete token (`freeze --force`) is a verb this CLI does not have.

    Each repair below names the engine that can actually satisfy its gate.
    The evidence gates name THIS one: ``_matching_validate_evidence`` accepts
    only evidence stamped ``python-hf-transformers`` and there is no
    run-substrate seam here, so evidence from the other engine can never
    satisfy them. Everything else is an authoring act, which is Mac-authority.
    """
    freeze_again = (f"steerlab-cli experiment freeze {name}  "
                    "(authoring is Mac-authority)")
    repairs = {
        "revision": f"steerlab-cli experiment create {name} --model <id> "
                    "--revision <commit> on the Mac, or load the model once "
                    f"here so it is cached, then {freeze_again}",
        "measurementPins": "repoint the invalid measurement pin at a loadable "
                           f"value on the Mac, then {freeze_again}",
        "validateEvidence": f"steerlab-server experiment validate {name}  "
                            "(this gate reads evidence stamped "
                            "python-hf-transformers; evidence from the other "
                            f"engine will not satisfy it), then {freeze_again}",
        "variantValidity": "re-save the variant with hashed adapter weights "
                           "and re-attach it on the Mac, then "
                           f"{freeze_again}",
        "batteryEvidence": f"steerlab-server experiment validate {name}  "
                           "(each variant condition runs the pinned battery), "
                           f"then {freeze_again}",
        "judgeValidity": f"steerlab-cli experiment pin-rubric {name} "
                         "prompts/rubrics/default-paired-v1.md --judges "
                         f"a:local[,b:claude] on the Mac, then {freeze_again}",
        "gitClean": "commit the pinned inputs in the workspace git repo, then "
                    f"{freeze_again}",
    }
    return repairs.get(
        gate, f"satisfy the '{gate}' gate, then {freeze_again}")


def admit_failures(name: str, failures: list[tuple[str, str]], *, force: bool) -> None:
    if failures and not force:
        raise ExperimentStoreError(
            failures[0][1], gate=failures[0][0],
            gates=tuple(g for g in FORCED_GATE_IDS if g in {gid for gid, _ in failures}),
            repair=freeze_gate_repair(failures[0][0], name))


def check_variant_validity(name: str, d: dict, *, evidence_grade_variants: set[int]) -> None:
    """Modality arms need a validity story (WORK-PLAN Phase E, promoted by the
    modality-axis decision): a variant condition's interventions must be fully
    pinned — adapter content hashes, system-prompt hash, vector artifact ids —
    or the condition is unverifiable the moment it freezes. ``freeze --force``
    skips this loudly, like the other evidence gates; the always-run verify()
    of the artifact-file hash is never skippable."""
    for index, vc in enumerate(d.get("variantConditions") or []):
        label = vc.get("name", "?")
        if isinstance(vc.get("fromPromotion"), dict):
            # Forward-referenced (stage 4): the artifact does not exist at
            # freeze time BY DESIGN — its pins land at run time and are
            # recorded in the run directory (forward-resolutions.json).
            # verify() enforces the declaration shape (attached concept,
            # exactly one identity).
            continue
        artifact = vc.get("artifact") or {}
        if not vc.get("artifactHash"):
            raise ExperimentStoreError(
                f"cannot freeze '{name}': variant '{label}' has no pinned "
                "artifactHash")
        for adapter in artifact.get("adapters") or []:
            if not adapter.get("adapterHash"):
                raise ExperimentStoreError(
                    f"cannot freeze '{name}': variant '{label}' adapter "
                    f"'{adapter.get('name', '?')}' has no adapterHash — re-save "
                    "the variant with hashed adapter weights, or freeze --force")
        if (artifact.get("systemPrompt") or "").strip() and \
                not artifact.get("systemPromptHash"):
            raise ExperimentStoreError(
                f"cannot freeze '{name}': variant '{label}' has a system prompt "
                "but no systemPromptHash — re-save the variant, or freeze --force")
        for injection in artifact.get("injections") or []:
            if not injection.get("vectorArtifactID"):
                raise ExperimentStoreError(
                    f"cannot freeze '{name}': variant '{label}' has an injection "
                    f"for '{injection.get('concept', '?')}' without a "
                    "vectorArtifactID pin")
        # Trained-adapter arms owe the same story about their TRAINING DATA
        # (LoRA readiness §0 amendment 1). An evidence-grade adapter whose
        # dataset is not pinned into the manifest is unverifiable the moment
        # it freezes: the training files could change afterwards with nothing
        # to flag the drift. Exploratory adapters are legal and produce an
        # advisory instead (freeze_advisories), never a refusal.
        if artifact.get("adapters") and index in evidence_grade_variants:
            block = vc.get("trainingProvenance")
            has_pin = isinstance(block, dict) and \
                str(block.get("datasetManifestHash") or "").strip()
            if not has_pin:
                raise ExperimentStoreError(
                    f"cannot freeze '{name}': variant '{label}' uses an "
                    "evidence-grade adapter but carries no "
                    "trainingProvenance.datasetManifestHash — its training "
                    "data would stay outside the freeze pin surface. Re-attach "
                    "the variant so freeze can pin the dataset manifest from "
                    "the adapter's sidecar, or freeze --force")


def check_battery_evidence(name: str, d: dict, evidence: dict | None,
                           *, expected_hash: str | None) -> None:
    """Variant freeze gate, evidence half: the scope-matched validate evidence
    must contain capability-battery results for the baseline and EVERY variant
    condition, produced from the battery the manifest pins (or the live
    default battery when unpinned). Hash pins alone say the artifact bytes are
    stable; this says the variant still answers concept-unrelated probes."""
    results = {r.get("condition"): r
               for r in (evidence or {}).get("batteryResults") or []
               if isinstance(r, dict)}
    # Forward-referenced conditions (stage 4) are exempt: their agent does
    # not exist at validate time, so their battery evidence is produced by
    # the RUN's per-condition battery, not by freeze-time validation.
    required = ["baseline"] + [
        vc.get("name", "?") for vc in d.get("variantConditions") or []
        if not isinstance(vc.get("fromPromotion"), dict)]
    missing = [c for c in required
               if c not in results or results[c].get("accuracy") is None]
    if missing:
        raise ExperimentStoreError(
            f"cannot freeze '{name}': validate evidence has no capability-"
            f"battery results for condition(s) {', '.join(missing)} — run "
            f"'experiment validate {name}' (each variant condition runs the "
            "pinned battery), or freeze --force")
    expected = expected_hash
    if expected:
        drifted = sorted(c for c in required
                         if results[c].get("batteryHash") != expected)
        if drifted:
            raise ExperimentStoreError(
                f"cannot freeze '{name}': capability battery drifted since "
                f"validation for condition(s) {', '.join(drifted)} — "
                f"re-run 'experiment validate {name}', or freeze --force")
